### PR TITLE
Add a parse_json_object function

### PR DESCRIPTION
**Pull request by NegativeMjark**
to deduplicate all the copy+pasted _parse_json functions. Also document
the parse_.* functions.

### CODE REVIEW

File: synapse/http/servlet.py [hunk 1]
```diff
@@ -60,6 +86,22 @@ def parse_boolean(request, name, default=None, required=False):
...
                  allowed_values=None, param_type="string"):
+    """Parse a string parameter from the request query string.
+
+    :param request: the twisted HTTP request.
+    :param name (str): the name of the query parameter.
+    :param default: value to use if the parameter is absent, defaults to None.
+    :param required (bool): whether to raise a 400 SynapseError if the
+        parameter is absent, defaulting to False.
+    :param allowed_values (list): List of allowed values for the string,
+        defaults to None
```
Comment (illicitonion) [inline synapse/http/servlet.py]: Presumably None means no restriction, rather than no allowed values?

File: synapse/http/servlet.py [hunk 2]
```diff
@@ -77,6 +119,23 @@ def parse_string(request, name, default=None, required=False,
...
+        SynapseError if the request body couldn't be decoded as JSON or
+            if it wasn't a JSON object.
+    """
+    try:
+        content = simplejson.loads(request.content.read())
+        if type(content) != dict:
+            raise SynapseError(400, "Content must be a JSON object.")
+        return content
+    except simplejson.JSONDecodeError:
+        raise SynapseError(400, "Content not JSON.")
```
Comment (illicitonion) [inline synapse/http/servlet.py]: No errcode?

File: synapse/http/servlet.py (hunk 1 above)
Comment (NegativeMjark) [inline synapse/http/servlet.py]: Done 2dba948

File: synapse/http/servlet.py (hunk 2 above)
Comment (NegativeMjark) [inline synapse/http/servlet.py]: Done  2dba948

**Comment by illicitonion** [comment]
LGTM

**Pull request merged**